core/src/main/scala/com/github/johnynek/paiges/DocTree.scala

```diff
@@ -0,0 +1,330 @@
+
+  private def space2(n: Int): Step[Nothing, Chunk] =
+    Emit(Str(" " * n)) // can memoize this
+
+  private def extract(s: String, part: String): Option[Str] =
+    if (s.startsWith(part)) Some(Str(s.substring(part.length)))
+    else None
+
+  /**
+   * Remove b from a
+   *
+   * consider None to be the empty set, which can't otherwise
+   * be represented
+   */
+  private[paiges] def setDiff(a: DocTree, b: DocTree): Option[DocTree] =
+    (a.unfix, b.unfix) match {
+      case (Split(as, bs) #:: _, _) =>
+        val adiff = setDiff(as, b)
+        def bdiff = setDiff(bs(), b)
+        adiff match {
+          case Some(ad) =>
+            bdiff match {
+              case Some(bd) => Some(docTree(Split(ad, () => bd) #:: Stream.empty))
+              case None => adiff
+            }
+          case None => bdiff
+        }
+      case (_, Split(as, bs) #:: _) =>
+        liftUnion(a) match {
+          case Split(x, y) =>
+            setDiff(docTree(Split(x, y) #:: Stream.empty), b)
+          case Emit(nounion) =>
+            setDiff(nounion, as) match {
+              case None => // we have already emptied, so we are done:
+                None
+              case Some(remainder) =>
+                setDiff(remainder, bs())
+            }
+        }
+      case (Stream.Empty, Stream.Empty) => None // this is now the empty set
+      case (Stream.Empty, _) => Some(a)
+      case (Emit(Break(nx)) #:: tailx, (right@Emit(Break(ny))) #:: taily) if (nx == ny) =>
+        setDiff(docTree(tailx), docTree(taily)).map { diff =>
+          docTree(right #:: diff.unfix)
+        }
+      case ((left@Emit(Str(s1))) #:: xtail, (right@Emit(Str(s2))) #:: ytail) =>
+        if (s1.length == s2.length) {
+          if (s1 == s2) {
+            setDiff(docTree(xtail), docTree(ytail)).map { diff =>
+              docTree(right #:: diff.unfix)
+            }
+          }
+          else Some(a)
+        }
+        else if (s1.length < s2.length) {
+          extract(s2, s1) match {
+            case Some(t) =>
+              setDiff(docTree(xtail), docTree(Emit(t) #:: ytail)).map { diff =>
+                docTree(left #:: diff.unfix)
+              }
+            case None => Some(a)
+          }
+        } else {
+          extract(s1, s2) match {
+            case Some(t) =>
+              setDiff(docTree(Emit(t) #:: xtail), docTree(ytail)).map { diff =>
+                docTree(right #:: diff.unfix)
+              }
+            case None => Some(a)
+          }
+        }
+      case (_, _) => Some(a) // they don't match at this point
+    }
+
+  /**
+   * does the docTree on the left contain the one on right
+   */
+  def isSubDoc(a: DocTree, b: DocTree): Boolean =
+    (a.unfix, b.unfix) match {
+      case (Split(as, bs) #:: _, _) =>
+        isSubDoc(as, b) && isSubDoc(bs(), b)
+      case (_, Split(as, bs) #:: _) =>
+        liftUnion(a) match {
+          case Split(aa, ab) =>
+            isSubDoc(docTree(Split(aa, ab) #:: Stream.empty), b)
```

Comment:
`def splitTree(a: DocTree, b: () => DocTree): DocTree = docTree(Split(a, b) #:: Stream.empty)` seems like it would come in handy in a few places.